main/at_transport_sdio.cpp

```diff
@@ -0,0 +1,390 @@
+    LOG(INFO, "Deinitializing SDIO transport");
+    if (rxThread_ || txThread_) {
+        exit_ = (rxThread_ != nullptr) + (txThread_ != nullptr);
+
+        LOG(INFO, "Waiting for SDIO threads to stop");
+
+        /* Join thread */
+        while (exit_) {
+            vTaskDelay(100 / portTICK_PERIOD_MS);
+        }
+
+        rxThread_ = nullptr;
+        txThread_ = nullptr;
+
+        LOG(INFO, "SDIO threads stopped");
+    }
+
+    waitWriteComplete(portMAX_DELAY);
+
+    started_ = false;
+
+    // Deinitialize SDIO pins
+    sdio_slave_deinit();
+
+    LOG(INFO, "SDIO transport deinitialized");
+
+    return 0;
+}
+
+int AtSdioTransport::postInitTransport() {
+    return 0;
+}
+
+int AtSdioTransport::fetchData(unsigned int timeoutMsec) {
+    sdio_slave_buf_handle_t handle;
+    size_t length = 0;
+    uint8_t* ptr = nullptr;
+
+    auto ret = sdio_slave_recv(&handle, &ptr, &length, timeoutMsec / portTICK_PERIOD_MS);
+    if (ret == ESP_ERR_TIMEOUT) {
+        return RESULT_TIMEOUT;
+    }
+    CHECK_ESP(ret);
+    Buffer* buf = container_of(ptr, Buffer, pbuf); // get struct list pointer
+
+    buf->handle = handle;
+    buf->leftLen = length;
+    buf->pos = 0;
+    buf->next = nullptr;
+
+    std::lock_guard<std::mutex> lock(rxMutex_);
+    if (!listTail_) {
+        listTail_ = buf;
+        listHead_ = listTail_;
+    } else {
+        listTail_->next = buf;
+        listTail_ = buf;
+    }
+
+    return length;
+}
+
+int AtSdioTransport::readData(uint8_t* data, ssize_t len, unsigned int timeoutMsec) {
+    if (data == nullptr || len < 0) {
+        return -1;
+    }
+
+    if (len == 0) {
+        return 0;
+    }
+
+    size_t toRead = len;
+    size_t pos = 0;
+    while (toRead > 0) {
+        Buffer* buf = (Buffer*)(listHead_);
+
+        // We don't have any more data
+        if (!buf) {
+            break;
+        }
+
+        ssize_t canRead = std::min<size_t>(buf->leftLen, toRead);
+        memcpy(data + pos, buf->pbuf + buf->pos, canRead);
+        buf->pos += canRead;
+        buf->leftLen -= canRead;
+        pos += canRead;
+        toRead -= canRead;
+
+        if (buf->leftLen == 0) {
+            // Can be given back
+            {
+                std::lock_guard<std::mutex> lock(rxMutex_);
+                listHead_ = buf->next;
+                buf->next = nullptr;
+
+                if (!listHead_) {
+                    listTail_ = nullptr;
+                }
+            }
+
+            auto ret = sdio_slave_recv_load_buf(buf->handle);
+            assert(ret == ESP_OK);
+
+            // Make sure to notify the host that we have a new buffer available
+            sdio_slave_send_host_int(HOST_SLC0_TOHOST_BIT0_INT_ENA_S);
+        }
+    }
+
+    rxData_ -= pos;
+
+    return pos;
+}
+
+int AtSdioTransport::flushInput() {
+    return 0;
+}
+
+int AtSdioTransport::writeData(const uint8_t* data, size_t len) {
+    if (len <= 0 || data == nullptr) {
+        return -1;
```

Comment:
Nit: magic number. Can we instead use `CHECK()` probably? Same to other sanity check.

Reply:
Just changed to `RESULT_ERROR`.